src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+from ..timer import Timer
+from ..widget import Widget
+from ..widgets import Label
+
+
+class Bar(Widget, can_focus=False):
+    """The bar portion of the progress bar."""
+
+    COMPONENT_CLASSES = {"bar--bar", "bar--complete", "bar--indeterminate"}
+    """
+    | Class | Description |
+    | :- | :- |
+    | `bar--bar` | Style of the bar (may be used to change the color). |
+    | `bar--complete` | Style of the bar when it's complete. |
+    | `bar--indeterminate` | Style of the bar when it's in an indeterminate state. |
+    """
+
+    DEFAULT_CSS = """
+    Bar {
+        width: 32;
+        height: 1;
+    }
+    Bar > .bar--bar {
+        color: $warning;
+        background: $foreground 10%;
+    }
+    Bar > .bar--indeterminate {
+        color: $error;
+        background: $foreground 10%;
+    }
+    Bar > .bar--complete {
+        color: $success;
+        background: $foreground 10%;
+    }
+    """
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of the bar that should be filled or `None` if indeterminate.
+
+    If the completion percentage is `None`, then the bar will render as
+    an indeterminate progress bar.
+    """
+    _refresh_timer: Timer | None
+    """Internal timer used to render the bar when it's in the indeterminate state."""
+
+    def watch__completion_percentage(self, percentage: float | None) -> None:
+        """Manage the timer that enables the indeterminate bar animation."""
+        if percentage is not None:
+            if self._refresh_timer:
+                self._refresh_timer.stop()
+                self._refresh_timer = None
+
+        else:
+            self._refresh_timer = self.set_interval(1 / 60, self.refresh)
+
+    def render(self) -> RenderResult:
+        """Render the bar with the correct portion filled."""
+        if self._completion_percentage is None:
+            return self.render_indeterminate()
+        else:
+            bar_style = (
+                self.get_component_rich_style("bar--bar")
+                if self._completion_percentage < 1
+                else self.get_component_rich_style("bar--complete")
+            )
+            return RenderableBar(
+                highlight_range=(0, self.size.width * self._completion_percentage),
+                highlight_style=Style.from_color(bar_style.color),
+                background_style=Style.from_color(bar_style.bgcolor),
+            )
+
+    def render_indeterminate(self) -> RenderResult:
+        """Render a frame of the indeterminate progress bar animation."""
+        width = self.size.width
+        highlighted_bar_width = 0.25 * width
+        # Width used to enable the visual effect of the bar going into the corners.
+        total_imaginary_width = width + highlighted_bar_width
+
+        speed = 30  # Cells per second.
+        # Compute the position of the bar.
+        start = (speed * monotonic()) % (2 * total_imaginary_width)
```

Comment:
I have feeling this may cause test flakiness in CI.

Suggest you have a method called `_get_time`. In tests you can monkey patch this.